File: csharp/perfect-numbers/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0)
        {
            throw new ArgumentOutOfRangeException($"{number} is not a natural number");
        }

        var factors = Factors(number);
        var aliquotSum = factors.Where(f => f <= number / 2).Sum();
        
        if (aliquotSum == number)
        {
            return Classification.Perfect;
        }

        return aliquotSum < number ? Classification.Deficient : Classification.Abundant;
    }

    private static IEnumerable<int> Factors(int number) => Enumerable.Range(1, number).Where(i => number % i == 0);
}

File: fsharp/perfect-numbers/PerfectNumbers.fs
module PerfectNumbers

type Classification =
    | Perfect
    | Abundant
    | Deficient

let classify n =
    let factors x =
        [ for i in 1 .. x / 2 do
              if x % i = 0 then yield i ]

    let aliquotSum x = List.sum (factors x)

    let classification =
        function
        | sum when sum < n -> Deficient
        | sum when sum > n -> Abundant
        | _ -> Perfect

    Some n
    |> Option.filter (fun x -> x > 0)
    |> Option.map (aliquotSum >> classification)